### DISCUSSION TITLE
Experimental Ansible JSON Schema for vscode

### DISCUSSION THREAD

**Discussion by ssbarnea** [ansible/ansible-lint#1399] [Show and tell] [answered]
I am currently working on generating Ansible JSON based validation schemas using a development branch of the linter and I need some feedback from users that are willing to help me.

At this moment activating these schemas is is done manually. After you install [YAML extension](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml), open your settings.json and add the following patterns:

```
  "yaml.schemas": {
    "https://raw.githubusercontent.com/ansible-community/ansible-lint/schemas/src/ansiblelint/f/ansible-requirements-schema.json": ["requirements.yml"],
    "https://raw.githubusercontent.com/ansible-community/ansible-lint/schemas/src/ansiblelint/f/ansible-meta-schema.json": ["meta/main.yml"],
    "https://raw.githubusercontent.com/ansible-community/ansible-lint/schemas/src/ansiblelint/f/ansible-vars-schema.json": ["vars/*.yml", "defaults/*.yml", "host_vars/*.yml", "group_vars/*.yml"],
  },
    "https://raw.githubusercontent.com/ansible-community/ansible-lint/schemas/src/ansiblelint/f/ansible-tasks-schema.json": ["tasks/*.yml", "handlers/*.yml"],
  },
    "https://raw.githubusercontent.com/ansible-community/ansible-lint/schemas/src/ansiblelint/f/ansible-playbook-schema.json": ["playbooks/*.yml"],
  },
```

Now when you open a file that matches these filenames, YAML extension will use these schemas to validate them. Obviously that they are quite incomplete.

Please remark that due to the dynamic nature of Ansible, we cannot fully make use of a centralized schema as each one would depend on currently installed version, modules, roles and collections. In fact the schema can differ ever between projects. The longer term goal is to make the vscode-ansible vscode extension to use the linter to generate and cache temporary schemas for each project but for the moment even generic shared schema should be ok.

**Comment by elreydetoda** [comment]
@ssbarnea, I tried to go to those urls now and it appears the files are no longer there. Am I missing something, did they move, or where they removed?

**Reply by elreydetoda** [reply]
as a follow up ( for others ), it appears they have all been moved here: https://github.com/ansible-community/schemas/tree/main/f

**Reply by ssbarnea** [reply]
Keep in mind that they are quite limited but I am more than happy to get some feedback from early adopters, maybe even some contributions.

**Reply by elreydetoda** [reply]
Sounds awesome! Thanks for maintaining these awesome things :grin:

**Reply by elreydetoda** [reply]
I am still new to this whole ansible playbook linting space ( your tools and stuff where mentioned [here](https://www.sherbers.de/the-better-ansible-vscode-extension/) ), but I will try to help out where I can! :slightly_smiling_face: 

I will apologize though, I am not the most knowledgeable about the topic :sweat_smile:

**Reply by g0t4** [reply]
I love the idea, giving it a try now!

**Comment by ssbarnea** [comment, marked as answer]
There were some major updated since this was proposed. We realised that JSON schema cannot be used to validate ansible playbooks/tasks due to insurmountable scalability issues. Still, other ansible file types can use schemas and we can use some bare minimal schemas for playbooks and tasks, but none that would be able to know about particular modules or their parameters.

The good news is that we already have a POC solution that is making use of language server at https://marketplace.visualstudio.com/items?itemName=redhat.ansible

**Reply by elreydetoda** [reply]
So, is it difficult to scale because of how JSON isn't very DRY, and that is making it not very possible since you have to keep repeating yourself in different areas just to keep thing maintained?

If that is the issue I have actually been looking at trying to use, and rather unsuccessfully ( probably haven't given it much time for me to wrap my head around everything ), @cue-lang (specifically [this repo](https://github.com/cue-lang/cue)). They say (IIRC) that it is a super-set of JSON, and in theory it should be able to make maintaining YAML & ( again IIRC ) JSON more easy, because you could just create and re-use functions to output whatever you need. I heard about it on [this podcast episode](https://changelog.com/gotime/163) by the @thechangelog. My idea was to use it to generate anisble playbooks, but I haven't had enough time to try and figure out how to use it yet :sweat_smile:

**Comment by bluikko** [comment]
Is the schema at https://raw.githubusercontent.com/ansible-community/schemas/main/f/ansible-lint.json out of date? It does not include `enable_list` for example.

**Reply by ssbarnea** [reply]
I fact I am actively working on that repo. Feel free to make a PR to add the missing entry.
